docs/rules/props-destructuring.md

```diff
@@ -0,0 +1,87 @@
+# Forbid multiline props destructuring at Component declaration (react/props-destructuring)
+
+Destructuring Component props on multiple line may confuse the reader where the actual block or the function start.
+
+## Rule Details
+
+This rule checks all React Component and verifies that props are not destructuring on multiple lines at component declaration. This rule is off by default.
+
+Examples of **incorrect** code for this rule:
+
+```jsx
+function Component({ 
+  prop1, 
+  prop2 
+}) {
+  return (
+    <div>
+      {prop1}
+      {prop2}
+    </div>
+  );
+}
+```
+
+```jsx
+const Component = ({ 
+  prop1, 
+  prop2 
+}) => (
+  <div>
+    {prop1}
+    {prop2}
+  </div>
+);
+```
+
+```jsx
+
+```
+
+```jsx
+
+```
+
+Examples of **correct** code for this rule:
+
+```jsx
+function Component(props) {
+  const { 
+    prop1, 
+    prop2 
+  } = props;
+  return (
```

Comment:
This rule in not against destructuring in the signature! It's agains destructuring __on multiple lines__  in the signature.
I agree that it's not the best example. 
A better one would be:
```jsx
function Component({ 
    prop1, 
    prop2,
    p3,
    p4,
    anotherProp: { deepDes },
    prop6,
    prop7,
}) {

  return (
    <div>
      {prop1}
      {prop2}
    </div>
  );
}
```
vs
```jsx
function Component(props) {
  const { 
    prop1, 
    prop2,
    p3,
    p4,
    anotherProp: { deepDes },
    prop6,
    prop7,
  } = props;

  return (
    <div>
      {prop1}
      {prop2}
    </div>
  );
}
```

__I with my team had times when there was hard to see from a glance where the component body starts.__

--- 

Also having more props on same line is ok, as it's still easy to see where the component starts. 
```jsx
function Component({prop1, prop2, p3, p4, prop6, prop7}) {

  return (
    <div>
      {prop1}
      {prop2}
    </div>
  );
}
```

---

I hope you now see a reason for this rule.

Reply:
I would still strongly argue that multi-line signature destructuring is the better pattern. The real issue there is that the component needs so many props :-)

It seems like syntax highlighting would help this case without needing to lose the declarative benefits of never reifying `props` as an object.

Reply:
> It seems like syntax highlighting would help this case without needing to lose the declarative benefits of never reifying props as an object.

What do you mean by this?

Reply:
> I would still strongly argue that multi-line signature destructuring is the better pattern.

I understand your point. Is it worth looking for a 3rd opinion?

Reply:
I'm sure we can get lots of opinions, but since the most popular styleguides, and most tutorials and docs, all mandate or recommend or demonstrate destructuring props in the signature, for half a decade now, and this is the first time this has come up - the implication is that it's a strikingly uncommon desire.

Reply:
Ok, I am closing this PR.